File: Obfuscation/RedSiege/12_v1.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Example shellcode (this would normally be read from a file)
unsigned char shellcode[] = {
    0xfc, 0x48, 0x83, 0xe4, 0xf0, 0xe8, 0xc0, 0x00,
    0x00, 0x00, 0x41, 0x51, 0x41, 0x50, 0x52, 0x51,
    0x56, 0x48, 0x31, 0xd2, 0x65, 0x48, 0x8b, 0x52,
    0x60, 0x48, 0x8b, 0x52, 0x18, 0x48, 0x8b, 0x52
};

#define SHELLCODE_SIZE sizeof(shellcode) / sizeof(shellcode[0])

// Function to generate a shuffled list of positions
void generate_shuffled_positions(int *positions, int size) {
    for (int i = 0; i < size; i++) {
        positions[i] = i;
    }

    // Shuffle the positions array using Fisher-Yates shuffle algorithm
    for (int i = size - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = positions[i];
        positions[i] = positions[j];
        positions[j] = temp;
    }
}

// Function to scramble the shellcode using the shuffled positions
void scramble_shellcode(unsigned char *jigsaw, int *positions, unsigned char *original, int size) {
    for (int i = 0; i < size; i++) {
        jigsaw[i] = original[positions[i]];
    }
}

// Function to reconstruct the original shellcode using the scrambled positions
void reconstruct_shellcode(unsigned char *reconstructed, int *positions, unsigned char *jigsaw, int size) {
    for (int i = 0; i < size; i++) {
        reconstructed[positions[i]] = jigsaw[i];
    }
}

// Function to print shellcode bytes in hex
void print_shellcode(unsigned char *code, int size) {
    for (int i = 0; i < size; i++) {
        printf("0x%02x, ", code[i]);
        if ((i + 1) % 8 == 0) {
            printf("\n");
        }
    }
    printf("\n");
}

int main() {
    srand(time(NULL)); // Seed the random number generator

    int positions[SHELLCODE_SIZE];
    unsigned char jigsaw[SHELLCODE_SIZE];
    unsigned char reconstructed[SHELLCODE_SIZE];

    // Generate shuffled positions
    generate_shuffled_positions(positions, SHELLCODE_SIZE);

    // Scramble the shellcode using the shuffled positions
    scramble_shellcode(jigsaw, positions, shellcode, SHELLCODE_SIZE);

    // Print original shellcode
    printf("Original Shellcode:\n");
    print_shellcode(shellcode, SHELLCODE_SIZE);

    // Print scrambled shellcode
    printf("\nScrambled Shellcode (Jigsaw):\n");
    print_shellcode(jigsaw, SHELLCODE_SIZE);

    // Reconstruct the original shellcode from the jigsaw array
    reconstruct_shellcode(reconstructed, positions, jigsaw, SHELLCODE_SIZE);

    // Print reconstructed shellcode
    printf("\nReconstructed Shellcode:\n");
    print_shellcode(reconstructed, SHELLCODE_SIZE);

    // Verify if the reconstruction was successful
    int success = 1;
    for (int i = 0; i < SHELLCODE_SIZE; i++) {
        if (shellcode[i] != reconstructed[i]) {
            success = 0;
            break;
        }
    }

    if (success) {
        printf("\nReconstruction successful!\n");
    } else {
        printf("\nReconstruction failed!\n");
    }

    return 0;
}

File: Obfuscation/RedSiege/12_v1.py
import random
import sys

# Function to read the shellcode from a binary file
def read_shellcode(filename):
    with open(filename, 'rb') as f:
        return list(f.read())

# Function to generate Jigsaw-encoded shellcode
def generate_jigsaw(shellcode):
    # Get the length of the shellcode
    shellcode_length = len(shellcode)

    # Create a list of positions corresponding to each byte in the shellcode
    raw_positions = list(range(shellcode_length))

    # Shuffle the positions to randomize the order
    random.shuffle(raw_positions)

    # Create the jigsaw-encoded shellcode by reordering the bytes
    jigsaw = [shellcode[i] for i in raw_positions]

    # Return the shuffled positions and the jigsaw shellcode
    return raw_positions, jigsaw

# Function to save the scrambled shellcode and positions to a file
def save_jigsaw_to_file(positions, jigsaw, output_file):
    with open(output_file, 'w') as f:
        # Writing the scrambled shellcode
        f.write("Jigsaw (scrambled shellcode):\n")
        f.write(', '.join(hex(byte) for byte in jigsaw) + '\n')

        # Writing the shuffled positions
        f.write("\nPositions (scrambled indices):\n")
        f.write(', '.join(str(pos) for pos in positions) + '\n')

# Function to reconstruct the original shellcode using the positions and jigsaw list
def reconstruct_shellcode(positions, jigsaw):
    shellcode_length = len(positions)
    reconstructed_shellcode = [0] * shellcode_length

    # Rebuild the shellcode by placing each byte at its original position
    for i, pos in enumerate(positions):
        reconstructed_shellcode[pos] = jigsaw[i]

    return reconstructed_shellcode

# Main function to tie everything together
def main(input_file, output_file):
    # Read the shellcode from the input binary file
    shellcode = read_shellcode(input_file)

    # Generate the jigsaw scrambled shellcode and positions
    positions, jigsaw = generate_jigsaw(shellcode)

    # Save the scrambled shellcode and positions to an output file
    save_jigsaw_to_file(positions, jigsaw, output_file)

    # Reconstruct the original shellcode
    reconstructed_shellcode = reconstruct_shellcode(positions, jigsaw)

    # Print the original and reconstructed shellcodes for comparison
    print("Original Shellcode:")
    print(', '.join(hex(byte) for byte in shellcode))
    print("\nReconstructed Shellcode:")
    print(', '.join(hex(byte) for byte in reconstructed_shellcode))

    # Check if the reconstruction was successful
    if shellcode == reconstructed_shellcode:
        print("\nReconstruction successful!")
    else:
        print("\nReconstruction failed!")

# Run the script
if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <input_shellcode_file> <output_file>")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    main(input_file, output_file)

